### PR TITLE
Adding support for forward configuration file

### PR DESCRIPTION
**Pull request by ajones**
### Description 
Support providing a configuration file for ip binding. It is passed via the `-z` or `--fwd-conf` flag.

### CODE REVIEW

**Review by theRemix — APPROVED** [review]
Works with and without the config option

New output looks great too! much easier to read the IP addrs in the same vertical column.

lgtm

**Comment by cjimti** [comment]
@ajones Thanks for the contribution. Starting looking at this and noticed a couple of issues.

1) `finalOctet` does not seem to make sense if you are forwarding multiple namespaces or clusters since the third octet increments on namespace and the second octet increments for each cluster. I would suggest simply making this a `reserveIP` and not allow automatic assignments to it, skipping this IP when incrementing. Also, I think this should be a reservation and not allow automatic assignment of other services to the reserved IP. As it is right now you could have other services assigned the same IP and if they use the same port they will fail to bind.
2) `serviceName` does not seem to take namespace or cluster into account. I have `postgres:5432` from the first namespace and `postgres.dev:5432` from my second namespace. Using the `serviceHostName` (as used in the logger) would be sufficient.
3) The formatting looks great. It's true that `30 chars is a pretty long service name` however please keep in mind that it is not just the service name if you account for statefulsets, namespaces and clusters.  Example: `web-keycloak-0.web-keycloak-headless.corp-internal:8080` and they get longer if multiple clusters are involved. I still think your 30 char column looks better than it did before.

-craig

**Comment by cjimti** [comment]
>  much easier to read the IP addrs in the same vertical column.

@theRemix I may be misunderstanding your comment, but the IP address have always been in a vertical column. 

@ajones It might be better to just format the IP column at 16 since we will never have an IP wider than 127.255.255.255 yet service names + (possible namespace) + (possible cluster name) is technically unbounded. 

Before
```
INFO[00:06:37] Port-Forward: 127.1.27.15 keycloak-2.keycloak-headless:8080 to pod keycloak-2:8080 
INFO[00:06:37] Port-Forward: 127.1.27.16 nifi:8443 to pod nifi-0:8443 
INFO[00:06:37] Port-Forward: 127.1.27.16 nifi:6007 to pod nifi-0:6007 
INFO[00:06:37] Port-Forward: 127.1.27.17 nifi-0.nifi:8443 to pod nifi-0:8443 
INFO[00:06:37] Port-Forward: 127.1.27.17 nifi-0.nifi:6007 to pod nifi-0:6007 
INFO[00:06:37] Port-Forward: 127.1.27.18 nifi-1.nifi:8443 to pod nifi-1:8443 
INFO[00:06:37] Port-Forward: 127.1.27.18 nifi-1.nifi:6007 to pod nifi-1:6007 
INFO[00:06:37] Port-Forward: 127.1.27.19 nifi-2.nifi:8443 to pod nifi-2:8443 
INFO[00:06:37] Port-Forward: 127.1.27.19 nifi-2.nifi:6007 to pod nifi-2:6007 
INFO[00:06:37] Port-Forward: 127.1.27.20 zk-hs:2888 to pod zk-0:2888 
INFO[00:06:37] Port-Forward: 127.1.27.20 zk-hs:3888 to pod zk-0:3888 
INFO[00:06:37] Port-Forward: 127.1.27.21 zk-0.zk-hs:2888 to pod zk-0:2888 
INFO[00:06:37] Port-Forward: 127.1.27.21 zk-0.zk-hs:3888 to pod zk-0:3888 
INFO[00:06:37] Port-Forward: 127.1.27.22 zk-1.zk-hs:2888 to pod zk-1:2888 
INFO[00:06:37] Port-Forward: 127.1.27.22 zk-1.zk-hs:3888 to pod zk-1:3888 
INFO[00:06:37] Port-Forward: 127.1.27.23 zk-2.zk-hs:2888 to pod zk-2:2888 
INFO[00:06:37] Port-Forward: 127.1.27.23 zk-2.zk-hs:3888 to pod zk-2:3888 
INFO[00:06:37] Port-Forward: 127.1.27.24 opsdb-cluster-repl:5432 to pod opsdb-cluster-1:5432 
INFO[00:06:37] Port-Forward: 127.1.27.25 superset-redis-master:6379 to pod superset-redis-master-0:6379 
INFO[00:06:37] Port-Forward: 127.1.27.26 nifi-registry:18080 to pod nifi-registry-664884cf69-9prtz:18080 
INFO[00:06:38] Port-Forward: 127.1.27.27 proxy-public:80 to pod proxy-559f856bd7-p2qjr:8000 
INFO[00:06:38] Port-Forward: 127.1.27.28 keycloak:8080 to pod keycloak-0:8080 
INFO[00:06:38] Port-Forward: 127.1.27.29 proxy-api:8001 to pod proxy-559f856bd7-p2qjr:8001 
INFO[00:06:38] Port-Forward: 127.1.27.30 nifi-port-proxy:8080 to pod nifi-port-proxy-75799c547c-kgfq4:8080 
INFO[00:06:38] Port-Forward: 127.1.27.31 pgadmin:5050 to pod pgadmin-0:5050 
INFO[00:06:39] Port-Forward: 127.1.27.32 nifi-ca-cs:8443 to pod nifi-ca-565d4559dc-z662g:8443 
INFO[00:06:39] Port-Forward: 127.1.27.33 zk-cs:2181 to pod zk-0:2181 
INFO[00:06:39] Port-Forward: 127.1.27.34 hub:8081 to pod hub-78fc6bbfd6-bdkhk:8081 
INFO[00:06:39] Port-Forward: 127.1.27.35 whdb-cluster-repl:5432 to pod whdb-cluster-1:5432 
```

After
```
INFO[23:59:57] Port-Forward: nifi-2.nifi:8443               127.1.27.15 to pod nifi-2:8443 
INFO[23:59:57] Port-Forward: nifi-2.nifi:6007               127.1.27.15 to pod nifi-2:6007 
INFO[23:59:57] Port-Forward: zk-hs:2888                     127.1.27.16 to pod zk-0:2888 
INFO[23:59:57] Port-Forward: zk-hs:3888                     127.1.27.16 to pod zk-0:3888 
INFO[23:59:57] Port-Forward: zk-0.zk-hs:2888                127.1.27.17 to pod zk-0:2888 
INFO[23:59:57] Port-Forward: zk-0.zk-hs:3888                127.1.27.17 to pod zk-0:3888 
INFO[23:59:57] Port-Forward: zk-1.zk-hs:2888                127.1.27.18 to pod zk-1:2888 
INFO[23:59:57] Port-Forward: zk-1.zk-hs:3888                127.1.27.18 to pod zk-1:3888 
INFO[23:59:57] Port-Forward: zk-2.zk-hs:2888                127.1.27.19 to pod zk-2:2888 
INFO[23:59:57] Port-Forward: zk-2.zk-hs:3888                127.1.27.19 to pod zk-2:3888 
INFO[23:59:57] Port-Forward: pgadmin:5050                   127.1.27.20 to pod pgadmin-0:5050 
INFO[23:59:57] Port-Forward: opsdb-cluster-repl:5432        127.1.27.21 to pod opsdb-cluster-1:5432 
INFO[23:59:57] Port-Forward: zk-cs:2181                     127.1.27.22 to pod zk-0:2181 
INFO[23:59:57] Port-Forward: proxy-api:8001                 127.1.27.23 to pod proxy-559f856bd7-p2qjr:8001 
INFO[23:59:58] Port-Forward: superset-redis-headless:6379   127.1.27.24 to pod superset-redis-master-0:6379 
INFO[23:59:58] Port-Forward: superset-redis-master-0.superset-redis-headless:6379 127.1.27.25 to pod superset-redis-master-0:6379 
INFO[23:59:58] Port-Forward: proxy-public:80                127.1.27.26 to pod proxy-559f856bd7-p2qjr:8000 
INFO[23:59:58] Port-Forward: keycloak:8080                  127.1.27.27 to pod keycloak-0:8080 
INFO[23:59:58] Port-Forward: whdb-cluster-repl:5432         127.1.27.28 to pod whdb-cluster-1:5432 
INFO[23:59:58] Port-Forward: hub:8081                       127.1.27.29 to pod hub-78fc6bbfd6-bdkhk:8081 
INFO[23:59:59] Port-Forward: keycloak-headless:8080         127.1.27.30 to pod keycloak-0:8080 
INFO[23:59:59] Port-Forward: keycloak-0.keycloak-headless:8080 127.1.27.31 to pod keycloak-0:8080 
INFO[23:59:59] Port-Forward: keycloak-1.keycloak-headless:8080 127.1.27.32 to pod keycloak-1:8080 
INFO[23:59:59] Port-Forward: keycloak-2.keycloak-headless:8080 127.1.27.33 to pod keycloak-2:8080 
INFO[23:59:59] Port-Forward: pmgsc-console:9443             127.1.27.34 to pod pmgsc-console-656f984594-bd85p:9443 
INFO[23:59:59] Port-Forward: nifi-port-proxy:8080           127.1.27.35 to pod nifi-port-proxy-75799c547c-kgfq4:8080 
```

**Comment by ajones** [comment]
@cjimti I sincerely appreciate the prompt feedback. I believe I have addressed your suggestions
1. added support for cluster and namespace identification in the conf file
2. detected ip/port collisions and allow the system to walk forward to find new option
3. cluster/namespace increment on octet 1 and 2 for wildcard matches
4. fixed width 16 on the `Port-Forward:` logline
5. panic guard on > 255 for an octet

I was able to test across 3 namespaces with identical pod deployments to assert the incrementing and reservations work as expected.

File: pkg/fwdservice/fwdservice.go [hunk 1]
```diff
@@ -282,7 +295,8 @@ func (svcFwd *ServiceFWD) LoopPodsToForward(pods []v1.Pod, includePodNameInHost
 				svcName,
 			)
 
-			log.Printf("Port-Forward: %s %s:%d to pod %s:%s\n",
+			// 30 chars is a pretty long service name
```
Comment (ajones) [inline pkg/fwdservice/fwdservice.go:298]:
```suggestion
```

**Comment by cjimti** [comment]
@ajones thanks for the updates! I have a work deadline tomorrow but can run some tests this weekend.

**Comment by ajones** [comment]
@cjimti sounds great, thanks for the heads up!

**Comment by ajones** [comment]
@cjimti I hope you had a great weekend. Would love to wrap this up so that we can begin using the change without forking. LMK if there are any outstanding issues you still wish for me to address.

**Comment by cjimti** [comment]
I have not had time to review it fully this weekend. Kubefwd does not have a set of mature tests or test automation. Any new features require quite a bit of manual testing, especially since most contributors do not use all the feature combinations and fail to test them (as happened earlier PR). Also, any new features to the codebase contribute to ongoing maintenance and testing demands, so I have to be thoughtful about their long-term consequences.

I appreciate your work on this and recommend that you create and use a custom release based on this PR, as it will probably be open for another week or two.

**Comment by ajones** [comment]
@cjimti thank you for the detailed response. I will do my best to address your recommendation. 

To your question about the use case. In our experience, there are many applications or tools that do not respect the entries in `/etc/hosts` when resolving a hostname. In particular, multiple PG tools like postico or mongo tools like Robo3t require the IP address to be provided to properly connect. Another situation is a react native application we previously developed is unable to resolve hostnames defined in `/etc/hosts` when running in the iOS simulator. 
In both of these situations, we are able to provide an IP address directly and everything works great. The issues come when the kubefwd process must be restarted due to forwarding failures. During this restart, the IPs change and then connection configurations must be manually updated.  
The failure mode for kubefwd here is when it is actively running and we destroy a deployment and replace it with another with the same name. This replacement is the side effect of a tool we use called garden.io which provides a convenient way to do development and testing in a remote k8 cluster. A replacement occurs each time we wish to deploy new code out to a service so you can imagine that the IPs changing every time we push up a change, to test would become repetitive to manage.
In summary, our end goal of all this is to have deterministic IP addresses for a set of pods running in a known namespace.

**Comment by cjimti** [comment]
The only way to have deterministic IP addresses would be to specify each service and the IP you want. Any other features like auto-incrementing would be non-deterministic since k8s does not guarantee namespace,  service, or pod order. New `kubefwd` runs would also get different IPs if other systems added removed services and Pods.

I think this could be specified on both the command-line or a config file, making it more flexible in automation.

Example:
`kubefwd svc -n ns1 -n ns2 --reserve a.ns1=127.0.0.5 --reserve a.ns2=127.1.27.15 --reserve b.ns1=10.1.1.0`

IPs in the reserve list are only assigned to the specified service/namespace/cluster.  With this method, it would also be easy to allow users to specify non-loopback IPs like (example `10.1.1.0`), a feature some have requested. 

When `kubefwd` loops through services and begins the IP assignment, it would check to see if the service has a reserved IP and use that, otherwise use its standard IP incrementor. Conversely, it would review the reservations before assigning a non-reserved service a reserved IP and skip to the next available IP.

Example with contexts (clusters)
```
serviceReservations:
  - hostname: a.ns1.svc.cluster.local
    ip: "127.0.0.5"
  - hostname: a.ns2.svc.cluster.local
    ip: "127.1.27.15"
  - hostname: b.ns1.svc.cluster.local
    ip: "10.1.1.0"
  - hostname: a.ns1.svc.cluster.prod-a
    ip: "127.1.1.15"
  - hostname: a.ns1.svc.cluster.prod-b
    ip: "127.1.27.8"
```

**Comment by ajones** [comment]
@cjimti it took me a moment to realize that the wildcard support was, as you point out, causing an issue. After taking another look at it I agree a full match should be made across ctx, ns, and service name to minimize these collisions. I have removed the wild card support so the system will utilize an IP provided in configuration if it can and if it can't, it logs an error and returns to incrementing behavior. I was previously not seeing this because I was ensuring there was not overlap from the 3rd octet and I was focused on deterministic IPs for the first targeted namespace.

To validate against your pg-a,b,c callout I deployed 3 pg services in the same topology as you outline and purposefully set their IPs to conflict with the 127.1.27.1,2,3 set and ensured when there was a collision with a reserved IP and one already taken the system was able to proceed.

**Comment by ajones** [comment]
@cjimti I hope you had a great weekend. I wanted to check in on that latest update I sent over and see if there were any further changes you would like to see made? I agree completely on the full match and fallback to the previous behavior if none is found.

**Comment by cjimti** [comment]
@ajones can you address my second post https://github.com/txn2/kubefwd/pull/208#issuecomment-930594773  I must have submitted it just before your reply so you may not have seen it.

**Comment by ajones** [comment]
Good evening @cjimti. Apologies if I am going to make you repeat yourself here, below I will attempt to outline how I see my changes addressing your requests. If you can help me find the gap I am happy to address it.

"IPs in the reserve list are only assigned to the specified service/namespace/cluster. With this method, it would also be easy to allow users to specify non-loopback IPs like (example 10.1.1.0), a feature some have requested."
I believe my previous change accomplishes this. Specifically here
https://github.com/txn2/kubefwd/pull/208/commits/95c391f810da0f8406e48dbbb7fd048b4bef1870#diff-0abe2083d683f1d1a09ac1a6a56de716a02a1c928bee6dde3a9ba225431f6d55L82
and here
https://github.com/txn2/kubefwd/pull/208/commits/95c391f810da0f8406e48dbbb7fd048b4bef1870#diff-0abe2083d683f1d1a09ac1a6a56de716a02a1c928bee6dde3a9ba225431f6d55R166
together those 2 lines force a full ctx/ns/service name match to apply a reserved IP.

"When kubefwd loops through services and begins the IP assignment, it would check to see if the service has a reserved IP and use that, otherwise use its standard IP incrementor."
I have confirmed my implementation located here
https://github.com/txn2/kubefwd/pull/208/files#diff-0abe2083d683f1d1a09ac1a6a56de716a02a1c928bee6dde3a9ba225431f6d55R80
will fall-through where there is not a perfect ctx/ns/svc match to here
https://github.com/txn2/kubefwd/pull/208/files#diff-0abe2083d683f1d1a09ac1a6a56de716a02a1c928bee6dde3a9ba225431f6d55R91
which is your previous implementation. 

"Conversely, it would review the reservations before assigning a non-reserved service a reserved IP and skip to the next available IP."
Here the system tries an egear IP assignment
https://github.com/txn2/kubefwd/pull/208/files#diff-0abe2083d683f1d1a09ac1a6a56de716a02a1c928bee6dde3a9ba225431f6d55R109
which checks the previously allocated addresses here
https://github.com/txn2/kubefwd/pull/208/files#diff-0abe2083d683f1d1a09ac1a6a56de716a02a1c928bee6dde3a9ba225431f6d55R121
before returning success/failure. In the case of failure the function will continue recursivly here
https://github.com/txn2/kubefwd/pull/208/files#diff-0abe2083d683f1d1a09ac1a6a56de716a02a1c928bee6dde3a9ba225431f6d55R114
which has the effect of incrementing here before attempting to allocate again
https://github.com/txn2/kubefwd/pull/208/files#diff-0abe2083d683f1d1a09ac1a6a56de716a02a1c928bee6dde3a9ba225431f6d55R108


In addition, to facilitate a better experience for command-line operation I will make the update to a single cluster.ns.svc in the conf file and support providing it by CLI.

**Comment by ajones** [comment]
@cjimti per your request for CLI provided reservations please see my latest commit. 
https://github.com/txn2/kubefwd/pull/208/commits/39d9cd7bbbd537f3201c30e1d05704af0ec0d030#diff-5c6d133a9ad0761c9ea9e58181c34d93d2da116a85cd62016d227a1c097d2f5eR78
If an identifier is provided in both the conf and cli flag the user receives a an info log of 
```
INFO[21:25:26] cli reservation flag overriding config for xxxxxx now 127.33.22.1 
```

as a note I wan unable to utilize the comma based syntax `foo.bar,127.x.y.z` because cobra would split these into the list instead of giving them to me as 1 str I could later work with. I opted for a `:` seperator instead.

**Comment by cjimti** [comment]
@ajones thank you for the detailed explanation. I have time to test tomorrow and Thursday. I appreciate you making these changes and look forward to getting this in the next release.

**Comment by ajones** [comment]
@cjimti no problem at all on timing, I understand how it goes. 
I do see your point and unfortunately in the x.y.z syntax (as you point out) there is ambiguity if x and y can be different things in different cases. To address this I see a few paths forward
1. do as you mention and add the pod-name component to the matcher list like this 
```
fmt.Sprintf("%s.%s.%s.svc.cluster.%s", o.PodName, o.ServiceName, o.Namespace, o.Context),
```
2. convert the yaml configuration back to multiple properties per block. e.g. namespace, service, etc... like I had it in the first pass. We would also need to change the cli formatting to be something like n:foo,s:bar,p:p1 for `pod-1.bar.foo.svc.cluster.local`. This is not my favorite as it adds a lot of noise and separation syntax to each entry but would resolve the ambiguity.

As a note, I think the confusion between [service].[namespace] and [pod-name].[service] will be helped by the fact the pod names come with rs and id information `foo-bar-68c77b448b-drjtx` whereas the service will not contain that same type of postfix information. This means at a quick glance I can distinguish between the purpose of these 2 
`foo-68c77b448b-drjtx.bar`
`bar.baz`

I went ahead and added support for pod-name because I felt it was the better of the 2 options. I look forward to your feedback.

**Comment by cjimti** [comment]
@ajones thanks, I'll pull a copy tonight. A lot of my headless services are backed by StatefulSets such as `db-node-0`,  `db-node-1`,  `db-node-2`, MinIO even prefixes it's own enumeration `obj-ss-0-0`, `obj-ss-0-1`, `obj-ss-0-2`. 

Keep in mind:
- You could have a service named `data` in the current namespace and access it with `data.svc`.
- You could also have the pod `db` in the namespace `data` and access it with  `db.data.svc`.

**Comment by cjimti** [comment]
@ajones should it check for reserved IPs before assigning? I assigned `-r ok:127.1.27.1` to a service but since that IP gets assigned first it was unable to assign it later on.



```
ERRO[18:29:24] Unable to forward service ok to requested IP 127.1.27.1 due to collision
```

**Comment by ajones** [comment]
@cjimti I'm on the same page with you. I will add a reservation conflict check in the `addToRegistry` function.

**Comment by ajones** [comment]
@cjimti I believe I have addressed the loopback and reservation check concerns. 
```
ERRO[14:40:15] ip 127.1.27.1 has already been allocated when placing XXX. will allocate next available 
ERRO[14:40:15] conflicting reservation for XXX on 127.1.27.2 when placing XXX. will allocate next available 
INFO[14:40:15] Port-Forward:       127.1.27.3 XXX to pod XXX
```
Also, I chose to panic for non-loopback reservations or `baseUnreservedIP`. Please let me know if that matches your expectation.
```
panic: IP 123.99.28.9 for XXX is not in the range 127.0.0.0/8
```

**Comment by cjimti** [comment]
@ajones great I can check it out later tonight!

IMO log.Fatal() would be more elegant than panic. I would make sense in a validation stage since it should be known immediately if any of the reserved IPs are not in the `127.0.0.0/8` range.


Example: https://stackoverflow.com/a/40072896/265026
